Make measurements resubmission faster

bassosimone opened the pull request:
# Description

This diff takes advantage of recents changes in mk* libraries [1] to implement faster measurements resubmission [2] for iOS.

We vendor mkcurl v0.12.0, mkbouncer v0.2.0, and a development version of mkcollector [3], to have code able to reuse existing open HTTP connections.

We implement a new MKReporter class that uses new code in mkcollector such that we open and close reports only when the measurements we are submitting are clearly part of different reports. That is, the app does not need to concern itself with this problem.

We do not anticipate specific issues in integrating with the OONI app for iOS because resubmission is sync code and therefore there are no significant concerns around keeping objects alive.

# Scope of the review

Please, ignore the files in `./vendor`. They are copied from other repositories where there is very high code coverage and they have also already been reviewed. Focus on the newly added ObjectiveC code and specifically on:

- [x] correctness
- [x] whether the API seems appropriate
- [x] whether more tests are needed

# Roll out plan

- [x] I will tag a stable version of mkall-ios out of this branch. *This means that the branch will qualify as being merged without squashing*. The reason why this operation is safe is because the only change is this new `MKReporter` interface and all the rest is unchanged. The release message will clarify that this interface is still unstable and subject to changes.

- [x] Then, I will create a branch in the https://github.com/ooni/probe-ios repository using this new version and I will make sure that we can smoothly transition to the new API.

- [x] When https://github.com/measurement-kit/mkcollector/pull/25 is approved, I will tag a new stable release of mkcollector and vendor that into this repository.

- [x] Once we have updated the probe-ios codebase, I will remove the previous interface for submitting measurements because at this point that has become obsolete.

- [x] This will open up the opportunity of upgrading this repository to MK v0.10.5, where a number of vendored files that where used by resubmission have now been removed.

# Notes

1. measurement-kit/measurement-kit#1867

2. ooni/probe#861

3. measurement-kit/mkcollector#25

bassosimone committed:
Make measurements resubmission faster

This diff takes advantage of recents changes in mk* libraries [1] to
implement faster measurements resubmission [2] for iOS.

We vendor mkcurl v0.12.0, mkbouncer v0.2.0, and a development
version of mkcollector [3], to have code able to reuse existing
open HTTP connections.

We implement a new MKReporter class that uses new code in mkcollector
such that we open and close reports only when the measurements we
are submitting are clearly part of different reports. That is, the
app does not need to concern itself with this problem.

We do not anticipate specific issues in integrating with the OONI app
for iOS because resubmission is sync code and therefore there are no
significant concerns around keeping objects alive.

.. [1] https://github.com/measurement-kit/measurement-kit/issues/1867

.. [2] https://github.com/ooni/probe/issues/861

.. [3] https://github.com/measurement-kit/mkcollector/pull/25

bassosimone commented on mkall/MKReporter.mm:
```diff
@@ -0,0 +1,93 @@
+MKUTIL_SET_INT(setTimeout, [](mk::collector::Settings *p, int64_t v) {
+  p->timeout = v;
+})
+
```
Here we also need to implement `deinit`.

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,107 @@
+                               softwareName:softwareName
+                            softwareVersion:softwareVersion
+  ];
+  return reporter;
```
This function can be written in a much more compact way

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,107 @@
+}
+
+// Utility function to create a measurement from a C string
+static NSString *makeMeasurement(const char *m) {
```
Maybe `makeSerializedMeasurement`?

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,107 @@
+
+@interface MKReporter : XCTestCase
+
+@end
```
The above interface should not be here

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,107 @@
+}
+
+// TODO(bassosimone): we need more complex tests involving submitting
+// several measurements within the same report.
```
This needs to be addressed before moving forward.

bassosimone reviewed (commented):
Pointed out places where more work is needed before others should review this code.

bassosimone commented on mkall/MKReporter.h:
```diff
@@ -0,0 +1,63 @@
+/// updatedSerializedMeasurement is the updated serialized measurement
+/// content where fields like the report ID have been updated taking into
+/// account the information that has been returned by the collector.
+@property NSString *updatedSerializedMeasurement;
```
We should write tests to make sure this is updated.

bassosimone commented on mkall/MKReporter.h:
```diff
@@ -0,0 +1,63 @@
+@property NSString *updatedSerializedMeasurement;
+
+/// updatedReportID is the updated report ID.
+@property NSString *updatedReportID;
```
We should write tests to make sure this is updated.

bassosimone commented on mkall/MKReporterInternal.h:
```diff
@@ -0,0 +1,25 @@
+#define XX(name_) unsigned name_;
+  MKCOLLECTOR_REPORTER_STATS_ENUM(XX)
+#undef XX
+};
```
Because this is now internal, we can probably use the mkcollector.hpp struct directly

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,180 @@
+
+@interface MKReporterTests : XCTestCase
+
+@end
```
Add `// @interface ...`

bassosimone commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,180 @@
+
+@implementation MKReporterTests
+
+// Utility function to submit just a single measurement.
```
Fix the docstring

bassosimone reviewed (commented):
More small changes required before blessing this as ready for review by @lorenzoPrimi

lorenzoPrimi commented on mkall/MKReporter.mm:
```diff
@@ -0,0 +1,95 @@
+      softwareName == nil || softwareVersion == nil) {
+    abort();
+  }
+  if (([self init]) != nil) {
```
Usually the conventional way to do this in Obj-C is
```
    self = [super init];
    if (self) {
//things
}
return self


```

bassosimone replied:
Will change the code, thanks for letting me know!

bassosimone committed:
Release v0.5.3

This is an interim release, functionally equivalent to v0.5.2 and based
off the same MK and dependencies, except that we're introducing a new API
that should allow for faster resubmissions.

See #25 and specifically the "Roll out plan" for details. More in detail,
this is step 1 of such plan.

bassosimone commented on mkall/MKReporter.h:
```diff
@@ -0,0 +1,63 @@
+
+/// setTimeout sets the optional per-request timeout. By default we
+/// don't set any timeout for submitting measurements.
+-(void)setTimeout:(int64_t)timeout;
```
Because the timeout is a per-request property, it's unfortunate to have it here. It would be much more convenient if the timeout was instead passed to `MKReporterTask:submit`.

bassosimone reviewed (commented):
More changes to be implemented to make this work correctly for iOS

bassosimone committed:
MK{Collector,Orchestra}.mm: fix initialisation bug

Spotted by @lorenzoPrimi, thanks!

bassosimone committed:
MKReporter: specify timeout when submitting

The timeout is a request dependent property because it is computed
using the body size in OONI for iOS.

This code pulls changes in https://github.com/measurement-kit/mkcollector/pull/25
to be able to specify the timeout when submitting.

bassosimone committed:
Relase v0.5.4

Like v0.5.4 except that I've adapted the `MKReporter` API such that
it is more easily useable from OONI for iOS.

bassosimone committed:
mkall.xcodeproj: enable diagnostics when debugging

bassosimone committed:
Make sure we actually free memory

This diff fixes a serious `facepalm.jpg` bug where the incorrect
method was overriden for releasing memory.

As such, this library was actually not releasing memory.

The `deinit` method is used by Swift. The `dealloc` method is
used by ObjectiveC. We should use `dealloc`.

It is surprising we never noticed. It may perhaps be because most
memory is managed by MK and only a little bit of it is instead
managed through ObjectiveC objects?

I enabled all sanitisers in the previous commit to increase our
confidence that we'll `facepalm.jpg` less in the future.

While there, do not declare methods such as `init` and `deinit`
COUGH COUGH `dealloc` that are supposedly already declared by
`NSObject`. This is an additional measure to making sure we are
really overriding, because otherwise Xcode complains.

I am also sad to note that is probably the most embarrasing
bug that I ever introduced (AFAIK, of course).

bassosimone committed:
Call `[self init]` in initialisers with arguments

This reverts 51290ad22394a98d3b39311510b3383e84a1ab35. I noticed
empirically that MKOrchestra doesn't work at all when we're calling
`[super init]` rather than `[self init]` in initializers that take
arguments. I do not currently have a precise explanation of why that
is the case. Intuitively, we want to call the initializer without
arguments of the current class, not of the parent class?

This, of course, does not imply that we should not call `[super init]`
when we're overriding the argument-less `init`.

bassosimone committed:
Release v0.5.5

This release fixes a very bad bug where we didn't free the memory
we allocated in ObjectiveC. It brings with it a bunch of stuff that
should make such major screwups less likely in the future.

As I mentioned in the related commit, it's interesting also to see
how we never noticed (or we never encountered data indicating the
occurrence of such event, perhaps).

bassosimone commented:
@lorenzoPrimi please take another look

lorenzoPrimi reviewed (approved):
Test it on the app as well

bassosimone commented:
> Test it on the app as well

Sure, please see: https://github.com/ooni/probe-ios/pull/302

bassosimone committed:
mkall: port code to MK v0.10.5

This basically means removing the `MKCollector` interface for which
MK support has been removed, adapting `MKGeoIPLookup` to directly use
C++ code, updating `MKReporter` to recent changes in `mkcollector`,
vendoring the most recent `mkcollector`, stop vendoring `mkcurl` and
use the version inside MK, and adapting `MKReporterTests` since we
cannot access `mkcurl.hpp` now from tests.

bassosimone committed:
Release v0.6.0

This release contains code changes required to use MK v0.10.5. It breaks
the API of the v0.5.x serious, so I cranked minor.

bassosimone commented:
I'm going to update ooni/probe-ios#302. @lorenzoPrimi, please take another look.

lorenzoPrimi commented on mkallTests/MKReporterTests.mm:
```diff
@@ -0,0 +1,231 @@
+// Ensure that we can submit a single measurement.
+- (void)testResubmissionGood {
+  NSDictionary *stats = @{
+    @"bouncer_error": @0,
```
shouldn't @ be only in front of string? I you want an int should be only
`@"bouncer_error": 0`, or am I missing something?

bassosimone replied:
The `@` in front of a number creates a `NSNumber`. I need to insert `NSNumbers` there mainly because that is what I do in the code that measures what happens.

lorenzoPrimi replied:
didn't know that, I always used `[NSNumber numberWithDouble:1000]`

lorenzoPrimi commented:
Approved
